File: go/pkg/zerotier/multicastgroup.go
/*
 * Copyright (c)2013-2020 ZeroTier, Inc.
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file in the project's root directory.
 *
 * Change Date: 2024-01-01
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by version 2.0 of the Apache License.
 */
/****/

package zerotier

import "fmt"

// MulticastGroup represents a normal Ethernet multicast or broadcast address plus 32 additional ZeroTier-specific bits
type MulticastGroup struct {
	MAC MAC    `json:"mac"`
	ADI uint32 `json:"adi"`
}

// String returns MAC#ADI
func (mg *MulticastGroup) String() string {
	if mg.ADI != 0 {
		return fmt.Sprintf("%s#%.8x", mg.MAC.String(), mg.ADI)
	}
	return mg.MAC.String()
}

// Less returns true if this MulticastGroup is less than another.
func (mg *MulticastGroup) Less(mg2 *MulticastGroup) bool {
	return mg.MAC < mg2.MAC || (mg.MAC == mg2.MAC && mg.ADI < mg2.ADI)
}

// key returns an array usable as a key for a map[]
func (mg *MulticastGroup) key() (k [2]uint64) {
	k[0] = uint64(mg.MAC)
	k[1] = uint64(mg.ADI)
	return
}

File: rust-zerotier-core/src/multicastgroup.rs
/*
 * Copyright (c)2013-2021 ZeroTier, Inc.
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file in the project's root directory.
 *
 * Change Date: 2026-01-01
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by version 2.0 of the Apache License.
 */
/****/

use crate::MAC;
use std::cmp::Ordering;
use std::hash::{Hash, Hasher};

#[derive(Clone, PartialEq, Eq)]
pub struct MulticastGroup {
    pub mac: MAC,
    pub adi: u32,
}

impl ToString for MulticastGroup {
    fn to_string(&self) -> String {
        format!("{}/{}", self.mac.to_string(), self.adi)
    }
}

impl Hash for MulticastGroup {
    #[inline(always)]
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.mac.0.hash(state);
        self.adi.hash(state);
    }
}

impl Ord for MulticastGroup {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.mac.0 < other.mac.0 {
            Ordering::Less
        } else if self.mac.0 > other.mac.0 {
            Ordering::Greater
        } else {
            if self.adi < other.adi {
                Ordering::Less
            } else if self.adi > other.adi {
                Ordering::Greater
            } else {
                Ordering::Equal
            }
        }
    }
}

impl PartialOrd for MulticastGroup {
    #[inline(always)]
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
